test: enable multiprocess tests on windows mac and ubuntu

zain-mecklai opened the pull request:
Now that we are able to successfully build these tests on the three desktop platforms, we need to start running them. This PR is part 2 of that process and removes the restriction to running the tests via automation.

zain-mecklai committed:
Start the process of getting multiprocess tests working on Windows and Mac

zain-mecklai committed:
test: enable multiprocess tests on Windows and mac

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Add Debugging to find who or what is null

zain-mecklai committed:
Merge branch 'chore/enable-multiprocess-tests-on-windows-mac' of github.com:Unity-Technologies/com.unity.netcode.gameobjects into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Build multiprocess tests on Windows and Mac

zain-mecklai committed:
Merge branch 'chore/enable-multiprocess-tests-on-windows-mac' of github.com:Unity-Technologies/com.unity.netcode.gameobjects into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Revert change to SnapshotSystem

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Remove tabs in document and replace with spaces

zain-mecklai committed:
Merge branch 'chore/enable-multiprocess-tests-on-windows-mac' of github.com:Unity-Technologies/com.unity.netcode.gameobjects into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Enable Linux Editor/Player tests for Multiprocess

NoelStephensUnity committed:
fix

fixes issues with clients being out of alignment with the server's scene in build list.
Includes client process command line argument to load the MultiprocessTestScene.
fixes issues with release MultiProcess client builds
fixes missing PrefabReference script component in MultiprocessTestScene's NetworkManager.

NoelStephensUnity committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

NoelStephensUnity committed:
fix

We should destroy the GameObject and not the component to avoid dependency issues.

NoelStephensUnity committed:
fix

MultiProcess tests no longer unload the TestRunner scene.
The server side MultiProcess scene is loaded additively and then set as the active scene until the OneTimeTearDown attribute decorated method TeardownSuite is invoked.
Added VerifySceneBeforeLoading callback handler to server side NetworkSceneManager in order to avoid sending the TestRunner scene as a scene that needs to be "synchronized" during client synchronization.
Adjusted the FixedUpdateMessagesAreOnlyProcessedOnceTest to not timeout so quickly as it was failing due to timing out too early.  This could have something to do with the last multiprocess client hanging around for 20 seconds before unloading.

NoelStephensUnity committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

NoelStephensUnity committed:
test-fix

Testing to see if just killing the processes at the end of a test suite will fix the issue where one client works but two clients fails.

NoelStephensUnity committed:
test-fix

Trying to determine where some of the issues are occurring when running in Yamato.
Adjusting when we create the processes as well as added additional debug information in the event that we run into the scenario where there is more than one TestCoordinator.

NoelStephensUnity committed:
test-fix

Trying this out to see how TestFixtures might be getting executed differently in Yamato than locally?  Several checks to see if we have already loaded the MultiProcessTests scene.

NoelStephensUnity committed:
test-fix

Excluding the NetworkVariablePerformanceTests for this run.
It has similar issues to the ExecuteStepInContext tests with the OneTimeSetup.

NoelStephensUnity committed:
test-fix

Removing the last TestCoordinatorTests TestFixture to determine if the MultiProcessTestScene is still hanging around when all MultiProcess tests are done or if it is related to the failure of the last 2 client TestCoordinatorTests TestFixture.

zain-mecklai committed:
Simplify logic on when to run teardown for multiprocess tests

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Revert changes

zain-mecklai committed:
Merge

NoelStephensUnity committed:
fix

This is the "bare minimum" adjustments to getting the multiprocess code base back to where it was running but having issues with TestFixtures.

zain-mecklai committed:
Enable Multiprocess tests

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Restore tests and try again

zain-mecklai committed:
codesign

zain-mecklai committed:
Fix job definition file so that it is valid

zain-mecklai committed:
disable graphics for the multiprocess tests

zain-mecklai committed:
Different args required for different platforms

zain-mecklai committed:
fix formatting

zain-mecklai committed:
Merge branch 'chore/enable-multiprocess-tests-on-windows-mac' of github.com:Unity-Technologies/com.unity.netcode.gameobjects into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Fix conditional syntax to validate yaml

zain-mecklai committed:
Revert changes as it isn't clear we need them

NoelStephensUnity committed:
Refactor

The base requirements to build and run multiprocess stuff.

zain-mecklai committed:
filter out the mobile platforms as the test doesn't run on them directly

NoelStephensUnity committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

NoelStephensUnity committed:
style

removing LF

NoelStephensUnity committed:
fix

Adding back the multiprocess ignore capabilities with the ability to easily override this ignore for continued development work.

NoelStephensUnity committed:
Merge branch 'develop' into test/multiprocess-fixes

NoelStephensUnity committed:
fix

including the IgnorMultiprocessTests test check for the NetworkVariablePerformanceTests

NoelStephensUnity committed:
style

fixing white space formatting issue due to indentation.

NoelStephensUnity committed:
Merge branch 'develop' into test/multiprocess-fixes

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Merge

zain-mecklai committed:
merge with develop

zain-mecklai committed:
Revert change back to original value

zain-mecklai committed:
Set display for ubuntu since we aren't using centos anymore

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
set ShouldIgnoreUTRTests to false always

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Revert changes back to what's on develop currently

zain-mecklai committed:
revert changes to what's on develop branch

zain-mecklai committed:
Fix standards check

zain-mecklai committed:
fix spacing issue in macro syntax

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
merge

zain-mecklai committed:
Copy and archive logfile from multiprocess tests

zain-mecklai committed:
revert to develop

zain-mecklai committed:
Turn off the multiprocess tests to get a clean baseline

zain-mecklai committed:
Try to run MultiprocessTests

zain-mecklai committed:
Continued iteration to find issue

zain-mecklai committed:
stop running multiprocess tests

zain-mecklai committed:
Don't copy files as they don't exist

zain-mecklai committed:
separate out multiprocess tests into separate job until we can get them stable

zain-mecklai committed:
Need Linq in order to use string Contains

zain-mecklai committed:
add editor version to yaml

zain-mecklai committed:
process logs after commands block

zain-mecklai committed:
not sure where the extra spaces come from in formatting the yaml but need to fix it

zain-mecklai committed:
starting to really hate yaml file formatting issues

zain-mecklai committed:
Add more diagnostics, update the flavor of mac to use the new standard b1.large value

zain-mecklai committed:
try to get more detailed logging and order of operations

zain-mecklai committed:
MAnage processes in a different way and log more carefully when processes are starting or not

zain-mecklai committed:
BuildRelease and also turn off graphics on Mac for faster operation

zain-mecklai committed:
Validate scripts

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
fix formatting and put back the UI

zain-mecklai committed:
fix missing script in scene and also only count active processes

zain-mecklai committed:
Try to get graphics showing on Linux

zain-mecklai committed:
Increase timeout slightly as some machines are slower than others

zain-mecklai committed:
Noels changes to get better stability and logging

zain-mecklai committed:
Merged with Noel's changes

NoelStephensUnity reviewed (approved):
Looks good Zain!

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/BaseMultiprocessTests.cs:
```diff
@@ -101,14 +101,19 @@ public virtual IEnumerator Setup()
             }
+            else
+            {
+                MultiProcessLog($"No need to spawn a new test player as there are already existing processes {MultiprocessOrchestration.ActiveWorkerCount()}");
```
thinking of it, how come we could already have workers running? Does this mean we aren't launching new players for each test fixture?
There's a danger with this, if tests don't properly cleanup their state, this could influence following runs. Especially for perf tests. Clearing players had the advantage of starting from a clean state for each fixture.
I understand we have resource contention. I'm guessing that change comes for this?

zain-mecklai replied:
I was trying to keep the logic as close to the original as possible. This was always how it was, clients were not disposed between test points and were reused across test points. I (along with @NoelStephensUnity ) tried to change the logic but ran into stability issues as the launch of a new client actually takes quite a while to process.

I'd actually prefer a clean, fresh client for each test point but the scope of that change is turning out to be quite large. We can tackle it here but I'd prefer to tackle that in a separate PR.

SamuelBellomo replied:
oh this is in setup. I think I misread and thought it was in SetupSuite.
Then, it's business as usual, ok, forget my comment

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/ExecuteStepInContext.cs:
```diff
@@ -56,7 +56,7 @@ public IEnumerator AfterTest(ITest test)
 
     private float m_StartTime;
-    private bool isTimingOut => Time.time - m_StartTime > TestCoordinator.MaxWaitTimeoutSec;
+    private bool isTimingOut => m_StartTime < Time.time;
```
won't this always return true?

SamuelBellomo replied:
oh just saw the line below. we might want to rename this var now, as it's not the "start time" anymore

NoelStephensUnity replied:
m_Timeout

zain-mecklai replied:
renamed variable

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/ExecuteStepInContextTests.cs:
```diff
@@ -142,31 +141,70 @@ public IEnumerator ContextTestWithAdditionalWait()
+                            sendNextResult = false;
+                        }
+
+                        // if we exceeded our maximum count, then remove ourselves from the update loop
```
```suggestion
                        // if we're finished sending our test count, then remove ourselves from the update loop
```

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/ExecuteStepInContextTests.cs:
```diff
@@ -142,31 +141,70 @@ public IEnumerator ContextTestWithAdditionalWait()
-                    TestCoordinator.Instance.WriteTestResultsServerRpc(count++);
-                    if (count > maxValue)
+                    if (!exitOnFailure)
                     {
-                        NetworkManager.Singleton.gameObject.GetComponent<CallbackComponent>().OnUpdate -= UpdateFunc;
+                        // We will send if we have another result to send =or= if we haven't received a response back from the server yet.
+                        if (sendNextResult || lastResultSentTime < Time.realtimeSinceStartup)
+                        {
+                            if (!sendNextResult && lastResultSentTime < Time.realtimeSinceStartup)
+                            {
+                                Debug.LogWarning($"{nameof(ContextTestWithAdditionalWait)} is resending due to timeout receiving result received response for count value = {count}!");
+                            }
+
+                            TestCoordinator.Instance.WriteTestResultsServerRpc(count);
+                            lastResultSentTime = resendResultTime + Time.realtimeSinceStartup;
+                            sendNextResult = false;
+                        }
```
I'm a bit confused by these lines.
If there's an RPC that fails sending, that should fail the test, there shouldn't be more retries, especially with a reliable RPC.
If that's needed, it'd be more useful to add these checks to ExecuteStepInContext, or even to TestCoordinator, so that all future multiprocess tests benefit from this boilerplate retry?

NoelStephensUnity replied:
This code could come out but was just there in the event that the server was missing messages.

SamuelBellomo replied:
That's my point above. If the server is missing messages, with reliable RPCs, that should fail the test.

SamuelBellomo replied:
If the server is missing messages, it means something is terribly wrong.
These basic tests are like Canaries. If they fail, it's very likely a lot of other tests will fail and that something is wrong at the core of Netcode.

NoelStephensUnity replied:
Exactly... but there was no type of check for that until now?
If you fail a test but have no reason as to why you failed the test...then how do you know why you failed the test?
But... it would mean bad things are happening indeed and as such remove it or leave it... both are fine IMO.

SamuelBellomo replied:
there was a check for that. With the nbFinished check below, you're are sure there was RPC exchange back and forth (since RPCs are used to trigger these actions).
Adding this check here just makes this test fail silently.

SamuelBellomo replied:
bump

SamuelBellomo replied:
We shouldn't have tests that should fail, pass because we're giving them too many chances.

zain-mecklai replied:
@NoelStephensUnity and @SamuelBellomo what should we do with this change

0xFA11 replied:
I'm inclined to agree with Sam here. Even if we can't immediately debug what was the issue causing tests to fail, I think we should still fail here instead of having retry attempts. Making tests fragile is generally a good idea instead of making them artificially robust by defensive checks, retrys and suppressions.
I'd vote to kill retry code even though I reckon it might make things harders to pass, which is a good thing IMO.

SamuelBellomo replied:
Just to get everyone on the same page, we went over these in private chat last week with Noel and agreed to remove these.
I think Zain is waiting for the SDK to stabilize a bit more (and for his return from vacation :P ) to continue work on this PR

SamuelBellomo replied:
@zain-mecklai I thought we agreed in our chat these would be removed? @JesseOlmer is saying you're ready to merge?

NoelStephensUnity replied:
@zain-mecklai
If you want me to assist with this I can as I was the one who added this code originally.

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -101,9 +102,9 @@ public void TestRunTeardown()
     }
 
-    public void OnDestroy()
+    public void OnDisable()
```
OnClientDisconnectCallback is added in Start()
Start is called only once.
For symmetry, you should also switch the subscription to this callback to OnEnable which can be called more than once, like OnDisable

zain-mecklai replied:
Please confirm that this is now correct

SamuelBellomo replied:
yep this is good

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -244,17 +245,24 @@ public void InvokeFromMethodActionRpc(Action methodInfo)
+            }
+            else
+            {
+                Instance.ClientFinishedServerRpc();
```
mmm so this is duplicating logic from step's Invoke, however, it ignores the logic about if (!m_WaitMultipleUpdates)
I'd forward ignoreException to invoke and move your if (!ignoreException) check to Invoke? what do you think?

NoelStephensUnity replied:
NO.
**Leave this here.**  It is the client side handling exceptions when you are explicitly testing for exceptions.
Otherwise, the client will have to timeout before it is considered "complete".  
This prevents that from happening.
If you are not ignoring/testing exceptions, then this will simply just "throw" as it did before.

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -310,18 +318,60 @@ public void KeepAliveClientRpc()
+    public KeepServerFromTimingOutDelegateHandler KeepServerFromTimingOut;
+
     [ServerRpc(RequireOwnership = false)]
+    //public void WriteTestResultsServerRpc(float result, ServerRpcParams receiveParams = default)
```
unused comment

NoelStephensUnity replied:
Yeah that needs to be removed.  It was an artifact.

zain-mecklai replied:
done

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -310,18 +318,60 @@ public void KeepAliveClientRpc()
+
+        // Now send the results received verification
+        m_TargetClient[0] = senderId;
+        m_ClientParams.Send.TargetClientIds = m_TargetClient;
```
mmm why is this a field and not just a var? I don't see it used anywhere else (and looking at the logic, it shouldn't be used anywhere else, as its value is based on the current context)

NoelStephensUnity replied:
m_ClientParams I would say we could make that a local var.
m_TargetClient should stay a field and pre-allocated.

zain-mecklai replied:
done, please take a look at it now and see if it is good to go.

SamuelBellomo replied:
I don't see any changes.
Noel: allocating an array of long of size 1 doesn't too big of a deal to me.
I'm nit picking here since there's a danger to having "global" right? m_ClientParams and m_Targetclient could be changed anywhere. If they are local to this context, it's self contained.

SamuelBellomo replied:
bump

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/ExecuteStepInContextTests.cs:
```diff
@@ -142,31 +141,70 @@ public IEnumerator ContextTestWithAdditionalWait()
+                    }
+                }
                 NetworkManager.Singleton.gameObject.GetComponent<CallbackComponent>().OnUpdate += UpdateFunc;
+                TestCoordinator.Instance.OnServerReceivedResultsResponse = TestResultReceived;
```
mmm I was about to tell you about yield return new WaitUntil(TestCoordinator.ResultIsSet());
but I'm remembering you can't yield in ExecuteStepInContext... that's a pain. And I think it could be doable to add this (however it'd be way out of scope of your current work :P )

NoelStephensUnity replied:
Yeah, all of that code could potentially be removed...it was all in the name of "sanity checking" while trying to figure out what was happening and when it was happening.
Or...we could leave that code there in the event we ever want to do verification that "data sent is data received".
There is merit behind double checking results.

SamuelBellomo replied:
there's already a way to do that though. That's what we're doing below server side with                         
Assert.That(res, Is.EqualTo(current++));

SamuelBellomo replied:
I'd say, let's remove this if it's not needed. Do we think checking if data sent is data received in a subsequent step is enough?

zain-mecklai replied:
@SamuelBellomo and @NoelStephensUnity is this still an open point of discussion? How do we close this?

SamuelBellomo replied:
As agreed in our offline discussion, this should be removed

NoelStephensUnity replied:
Yes. we agreed that this should be removed too.

zain-mecklai committed:
Update testproject/Assets/Tests/Runtime/MultiprocessRuntime/ExecuteStepInContextTests.cs

Co-authored-by: Sam Bellomo <71790295+SamuelBellomo@users.noreply.github.com>

zain-mecklai committed:
Shut down processes in between tests

zain-mecklai committed:
Shut down processes in between tests

zain-mecklai committed:
Fix whitespace formatting

zain-mecklai committed:
in TearDown close all processes we spawned

NoelStephensUnity commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/BaseMultiprocessTests.cs:
```diff
@@ -141,6 +140,8 @@ public virtual void Teardown()
             if (!IgnoreMultiprocessTests)
             {
+                MultiProcessLog($"Stop all workers after each test so we can start fresh ones.");
+                MultiprocessOrchestration.ShutdownAllProcesses();
```
Remove this from the TearDown completely.
This is already done in this method:
    [OneTimeTearDown]
    public virtual void TeardownSuite()

zain-mecklai replied:
Yup, removed.

zain-mecklai committed:
Try to recycle spawned processes between tests

NoelStephensUnity commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -22,7 +23,7 @@ public class TestCoordinator : NetworkBehaviour
     public const int PerTestTimeoutSec = 5 * 60; // seconds
 
-    public const float MaxWaitTimeoutSec = 20;
+    public const float MaxWaitTimeoutSec = 60;
```
We could move this back down to like 30 seconds if we wanted...

zain-mecklai replied:
okie

zain-mecklai replied:
Done

zain-mecklai committed:
PR feedback

zain-mecklai committed:
There seems to be a small chance that an existing process dies while we are setting up, confirm this and recover if possible.

zain-mecklai committed:
Workaround for UnityLog.txt now being archived for debugging

zain-mecklai committed:
Continue to dig into Mac failures

zain-mecklai committed:
mid process

zain-mecklai committed:
try to incrementally increase the count

zain-mecklai committed:
Continue to diagnose

zain-mecklai committed:
revert back to weekend version of code and redo each change slowly

zain-mecklai committed:
Might need a more powerful mac for all the players that are running

zain-mecklai committed:
Roll back

zain-mecklai committed:
More logging

zain-mecklai committed:
recert again

zain-mecklai committed:
Wait until the async operation of unloading the scene is done before moving on

zain-mecklai committed:
Revert changes to try and yield for the scene unload to take place, instead just report on it

zain-mecklai committed:
add more diagnostics and hope that it shows something useful on failure

zain-mecklai committed:
Try to figure out why Mac is having so much trouble

zain-mecklai committed:
added diagnostics

zain-mecklai committed:
added more diagnostics and polling

zain-mecklai committed:
Catch NullReferenceException in client for NUnit Test Name

zain-mecklai committed:
Try restarting processes if we can't get the connected client count to where we need

zain-mecklai committed:
Try starting an additional worker if we haven't gotten our connection count up to par yet

zain-mecklai committed:
PR Feedback

zain-mecklai committed:
move callback set up to OnEnable to match OnDisable removal

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
Fix whitespace

zain-mecklai committed:
redo start up logic for workers

zain-mecklai committed:
Add diagnostic about the log file name

zain-mecklai committed:
Increased logging and attempt to capture screenshot

zain-mecklai committed:
Continue to iterate on getting deeper diagnostic information on failure

zain-mecklai committed:
Update logging messages and try to run with no graphics to see if that makes a difference

zain-mecklai committed:
Run with no graphics on linux as well and try to fix whitespace standards

zain-mecklai committed:
static initialize class to make sure directory info data is set

zain-mecklai committed:
formatting and remove copy command for pngs since we no longer have graphics

zain-mecklai committed:
fix whitespace formatting

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -52,16 +53,21 @@ private void Awake()
         bool isClient = Environment.GetCommandLineArgs().Any(value => value == MultiprocessOrchestration.IsWorkerArg);
         if (isClient)
         {
+            NetworkManager.NetworkConfig.EnableNetworkLogs = true;
```
do we want this for all multiprocess tests? logs will slow down perf tests for example

zain-mecklai replied:
okay, I'll remove it, I was just trying to figure out why sometimes connections never established. Perhaps we can put it back when we actually have a problem. Where do these logs go anyhow?

zain-mecklai replied:
addressed in https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1195/commits/4d26c46c0d7b955b296b3162b28c995fdd2c45b3

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/BaseMultiprocessTests.cs:
```diff
@@ -109,18 +109,36 @@ public virtual IEnumerator Setup()
             {
-                var numProcessesToCreate = WorkerCount - MultiprocessOrchestration.ActiveWorkerCount();
-                for (int i = 0; i < numProcessesToCreate; i++)
+                var timeOutTime2 = Time.realtimeSinceStartup + TestCoordinator.MaxWaitTimeoutSec / 3;
```
/ 3? Where's that coming from?
If it's because MaxWaitTimeoutSec is too long, why not just have a separate one for this case?

zain-mecklai replied:
Actually, now that I'm looking at it, I was intending to do some extra processing to start a new worker if the previous worker somehow didn't start up but that's a really bad idea and would hide a legit failure in the case that the worker failed to start. I'll revert this change.

zain-mecklai replied:
addressed in https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1195/commits/4d26c46c0d7b955b296b3162b28c995fdd2c45b3

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/Helpers/MultiprocessOrchestration.cs:
```diff
@@ -16,6 +16,11 @@ public class MultiprocessOrchestration
 
+    static MultiprocessOrchestration()
+    {
+        GetPathToMultiprocessDirectory();
```
A "get" is usually idempotent, this should be called "Initialize" or something like that?

zain-mecklai replied:
that's a fair point. I shall rename it.

zain-mecklai replied:
https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1195/commits/59f7306dc56222c15fc7e05d95c0a65164ea55a9

zain-mecklai committed:
static initialize the multiprocess directory infot

zain-mecklai committed:
Remove extra polling at 1/3rd time. Also make local var where a field or property was not needed.

zain-mecklai committed:
PR feedback

zain-mecklai committed:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

zain-mecklai committed:
add to run-all macro

zain-mecklai committed:
add to run-all macro and fix type

zain-mecklai committed:
Revert some additional debugging steps

zain-mecklai committed:
merge conflicts

zain-mecklai committed:
Print null reference exception

zain-mecklai committed:
try to remove GPU requirement, clean up whitespace

0xFA11 commented on .yamato/project.metafile:
```diff
@@ -8,16 +8,19 @@ test_platforms:
     type: Unity::VM::osx
     image: package-ci/mac:stable
-    flavor: m1.mac
+    flavor: b1.xlarge
```
`large` as others vs `xlarge`?

zain-mecklai replied:
Good point, initially I thought that the xlarge would solve the startup timing issues but that was not the case at all, I shall revert this.

zain-mecklai replied:
Please take a look at [this change](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1195/commits/c8004bed8f4b89feec07653aa72b9be94ef997dd) and mark conversation as resolved if you are okay with it.

0xFA11 commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -246,17 +259,24 @@ public void InvokeFromMethodActionRpc(Action methodInfo)
 
     [ClientRpc]
-    public void TriggerActionIdClientRpc(string actionId, byte[] args, ClientRpcParams clientRpcParams = default)
+    public void TriggerActionIdClientRpc(string actionId, byte[] args, bool ignorException, ClientRpcParams clientRpcParams = default)
```
nit :P
```suggestion
    public void TriggerActionIdClientRpc(string actionId, byte[] args, bool ignoreException, ClientRpcParams clientRpcParams = default)
```

zain-mecklai replied:
Please take a look at [this change](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1195/commits/c8004bed8f4b89feec07653aa72b9be94ef997dd) and mark conversation as resolved if you are okay with it.

JesseOlmer commented on .yamato/project-tests.yml:
```diff
@@ -54,3 +54,38 @@ test_{{ project.name }}_{{ editor }}_{{ platform.name }}:
+{% for editor in project.test_editors -%}
+{% for platform in test_platforms -%}
+multiprocess_test_testproject_{{ editor }}_{{ platform.name }}:
+  name : testproject multiprocess project tests - {{ editor }} on {{ platform.name }}
```
nit: these names are getting super long in the UI - `multiprocess tests - {{ editor }} on {{ platform.name }}` conveys the same info and saves us 20 characters, hopefully meaning the variable bits show up on the screen.

zain-mecklai replied:
That's totally fair, doing this now.

zain-mecklai replied:
Please take a look at the change and resolve this conversation if you feel it is addressed.

zain-mecklai replied:
Please take a look at [this change](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1195/commits/c8004bed8f4b89feec07653aa72b9be94ef997dd) and mark conversation as resolved if you are okay with it. @JesseOlmer

JesseOlmer commented on .yamato/project-tests.yml:
```diff
@@ -54,3 +54,38 @@ test_{{ project.name }}_{{ editor }}_{{ platform.name }}:
+    - pip install unity-downloader-cli --index-url https://artifactory.prd.it.unity3d.com/artifactory/api/pypi/pypi/simple
+    - unity-downloader-cli -u {{ editor }} -c editor -w --fast
+    - {% if platform.name != "win" %}curl -s https://artifactory.prd.it.unity3d.com/artifactory/unity-tools-local/utr-standalone/utr --output utr && chmod +x ./utr{% endif %}{% if platform.name == "win" %}curl -s https://artifactory.prd.it.unity3d.com/artifactory/unity-tools-local/utr-standalone/utr.bat --output utr.bat{% endif %}
+    - {% if platform.name == "ubuntu" %}DISPLAY=:0 && {% endif %}{{ platform.editorpath }} -projectpath testproject -batchmode -nographics -quit -logfile BuildMultiprocessTestPlayer.log -executeMethod Unity.Netcode.MultiprocessRuntimeTests.BuildMultiprocessTestPlayer.BuildRelease
```
`{% if platform.name == "ubuntu" %}DISPLAY=:0 && {% endif %}`
`DISPLAY=:0` sets a variable scoped to the given command, and then `&&` "starts" a new command... meaning your display variable is NOT applied to the editor command that follows. Looks like you don't need this `DISPLAY=:0` at all (makes sense with batchmode and nographics)

JesseOlmer commented on .yamato/project-tests.yml:
```diff
@@ -54,3 +54,38 @@ test_{{ project.name }}_{{ editor }}_{{ platform.name }}:
+  commands:
+    - pip install unity-downloader-cli --index-url https://artifactory.prd.it.unity3d.com/artifactory/api/pypi/pypi/simple
+    - unity-downloader-cli -u {{ editor }} -c editor -w --fast
+    - {% if platform.name != "win" %}curl -s https://artifactory.prd.it.unity3d.com/artifactory/unity-tools-local/utr-standalone/utr --output utr && chmod +x ./utr{% endif %}{% if platform.name == "win" %}curl -s https://artifactory.prd.it.unity3d.com/artifactory/unity-tools-local/utr-standalone/utr.bat --output utr.bat{% endif %}
```
nit: you've got 2 entirely different commands on this 1 line (windows and non-windows) which makes it pretty hard to read. Why not do it the way you do the codesigning command below and put windows command on a separate line/instruction than non-windows to make this a bit more readable?

JesseOlmer commented on .yamato/project-tests.yml:
```diff
@@ -54,3 +54,38 @@ test_{{ project.name }}_{{ editor }}_{{ platform.name }}:
+{% if platform.name == "mac" %}    -  sudo codesign --force --deep --sign - ./testproject/Builds/MultiprocessTests/MultiprocessTestPlayer.app{% endif %}
+    - {% if platform.name == "ubuntu" %}DISPLAY=:0 && {% endif %}{{ platform.utr }} --suite=playmode --testproject=testproject --editor-location=.Editor --testfilter=Unity.Netcode.MultiprocessRuntimeTests --extra-editor-arg=-bypassIgnoreUTR
+  after:
+    - {% if platform.name == "win" %}copy %USERPROFILE%\.multiprocess\logfile* .{% endif %}{% if platform.name != "win" %}cp $HOME/.multiprocess/logfile* .{% endif %}
```
nit: same here with the hard-to-read multiple os commands on a single line format

JesseOlmer commented on .yamato/project-tests.yml:
```diff
@@ -54,3 +54,38 @@ test_{{ project.name }}_{{ editor }}_{{ platform.name }}:
+        - BuildMultiprocessTestPlayer.log
+        - "logfile*"
+        - "*.log"
+        - "*.png"
```
why are we saving pngs?

zain-mecklai replied:
It was to try and identify the reason for some oof the failures we were seeing (I was hoping to capture some visual artifact) but since this has all gone to batchmode with nographics this is no longer needed, I will remove.

JesseOlmer commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/BaseMultiprocessTests.cs:
```diff
@@ -95,20 +99,41 @@ private bool VerifySceneIsValidForClientsToLoad(int sceneIndex, string sceneName
+                    MultiProcessLog($"Spawning testplayer {i} since connected client count is {NetworkManager.Singleton.ConnectedClients.Count} is less than {WorkerCount} and Number of spawned external players is {MultiprocessOrchestration.ActiveWorkerCount()} ");
+                    string logPath = MultiprocessOrchestration.StartWorkerNode(); // will automatically start built player as clients
+                    MultiProcessLog($"logPath to new process is {logPath}");
+                    yield return new WaitForSeconds(5.0f); // Wait 5 seconds before starting the next process because there are odd contentions that happen when starting multiple processes quickly
```
boo :( our 10-player tests will take over a minute to even start executing... oh well :/

zain-mecklai replied:
Let me take another look at this, I didn't need it before and perhaps it is just a symptom.

zain-mecklai replied:
Dropping this to 0.1f seconds.

JesseOlmer commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/Helpers/MultiprocessOrchestration.cs:
```diff
@@ -91,37 +134,37 @@ public static void StartWorkerNode()
         }
         catch (Win32Exception e)
         {
             Debug.LogError($"Error starting player, {buildInstructions}, {e.Message} {e.Data} {e.ErrorCode}");
```
MultiProcessLog?

JesseOlmer commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -39,6 +40,7 @@ public class TestCoordinator : NetworkBehaviour
+        BaseMultiprocessTests.MultiProcessLog("Awake");
         if (Instance != null)
         {
             Debug.LogError("Multiple test coordinator, destroying this instance");
```
MultiProcessLog?

zain-mecklai replied:
Yes, done.

JesseOlmer commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/BaseMultiprocessTests.cs:
```diff
@@ -135,17 +163,45 @@ public virtual void Teardown()
+            }
+        }
+
+        public static void MultiProcessLog(string msg)
```
nit: https://docs.unity3d.com/ScriptReference/Logger.html

zain-mecklai replied:
I'm not clear on how to use this to control the logging format just yet.

zain-mecklai commented:
Please allow me a little time to respond to all the comments. @JesseOlmer  @MFatihMAR @SamuelBellomo

SamuelBellomo commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -312,21 +332,54 @@ public void KeepAliveClientRpc()
+    public delegate void OnServerReceivedResultsResponseDelegateHandler(float resultReceived);
+
+    public OnServerReceivedResultsResponseDelegateHandler OnServerReceivedResultsResponse;
+    private void ServerReceivedResultsResponse(float resultReceived)
```
same here, as agreed this should be removed

zain-mecklai committed:
reduce size of mac vm image, fix spelling error and rename build step to be shorter

zain-mecklai committed:
simplify and separate out platform specific commands in yaml files, remove unneeded diagnostic code from TestCoordinator

zain-mecklai committed:
make smaller changes

zain-mecklai committed:
Remove extra diagnostic and debugging code

zain-mecklai committed:
Revert changes

zain-mecklai committed:
Attempt to use Unity's own loggers

zain-mecklai committed:
clean up whitespace and try to use logging more consistently

zain-mecklai committed:
set the ignore exception state correctly

NoelStephensUnity committed:
fix

Adding the ability for MultiProcessLog to have warnings and errors as well as the ability to bypass formatting so when you are trying to trap for a specific exception message you don't have to include the additional MultiProcessLog overhead text.

NoelStephensUnity committed:
refactor

Using the ignoreException so only when we want to test for exceptions do we ignore the formatting.

NoelStephensUnity reviewed (approved):
Looks really nice Zain!

zain-mecklai committed:
whittle down the random delay in starting processes

zain-mecklai committed:
Log error messagees using the neew logging framework that utilizes unity's built in logging

zain-mecklai committed:
use new logging and drop wait time aftr spawning process to .1 seconds

JesseOlmer commented on testproject/Assets/Tests/Runtime/MultiprocessRuntime/TestCoordinator.cs:
```diff
@@ -325,8 +345,16 @@ public void WriteTestResultsServerRpc(float result, ServerRpcParams receiveParam
 
     [ServerRpc(RequireOwnership = false)]
-    public void WriteErrorServerRpc(string errorMessage, ServerRpcParams receiveParams = default)
+    public void WriteErrorServerRpc(string errorMessage, bool ignoreFormatting = false, ServerRpcParams receiveParams = default)
```
@SamuelBellomo @NoelStephensUnity This parameter was added here, affects the message formatting in this method body, and controls whether the message is further formatted with "MPLOG {Date}" as it's passed along to `BaseMultiprocessTests.MultiProcessLog` or if it instead looks like a regular non-multiprocess log... **HOWEVER**, the only caller to this method which sets this boolean is `TriggerActionIdClientRpc`, where the boolean is NOT called `ignoreFormatting` but is instead called `ignoreException`. This seems like a usage mismatch, and that we don't even need this boolean here.

As if that wasn't confusing enough, the only place `ignoreException` in `TriggerActionIdClientRpc` is called, the boolean is actually called `m_IgnoreTimeoutException`, but `TriggerActionIdClientRpc` treats it as a catch-all for ANY `Exception`

To summarize: `IgnoreTimeoutException`->`ignoreException`->`catch(Exception)`->`ignoreFormatting`

long and short of it is... I have no clue what the intention is here of this stuff as a single boolean is used in 4 different places, called 3 different things, for 4 seemingly different purposes. Can someone please clarify what this should actually do?

P.S. THIS IS WHY I HATE BOOLEAN AND DEFAULT PARAMS!

zain-mecklai replied:
@JesseOlmer do you want me to take a crack at resolving this issue? I think this is the last issue pending for this PR.

JesseOlmer replied:
Already fixed and merged into this branch/PR - apologies for leaving this open.

NoelStephensUnity replied:
It was a bad implementation on my part.

zain-mecklai committed:
Remove delay in starting processes

Commit:
test: refactor multiprocesslogger (#1254)

Move MultiProcessLog formatting into MultiprocessLogHandler
Remove special exception logging from multiprocess

Commit:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac

Commit:
Merge branch 'develop' into chore/enable-multiprocess-tests-on-windows-mac
